ethereum/contracts/utils/ICM.sol

```diff
@@ -0,0 +1,377 @@
+            serialized[14 + addressedCall.sourceAddress.length + i] = addressedCall.payload[i];
+        }
+
+        return serialized;
+    }
+
+    function filterValidators(
+        bool[] memory signers,
+        Validator[] memory validators
+    ) internal view returns (bytes memory, uint64) {
+        require(
+            signers.length <= validators.length,
+            "Signers length must be less than or equal to validators length"
+        );
+        bytes memory aggregatePublicKey;
+        uint64 aggregateWeight = 0;
+        for (uint256 i = 0; i < validators.length; ++i) {
+            if (i < signers.length && signers[signers.length - i - 1]) {
+                // Cache the validator to avoid repeated array access
+                Validator memory validator = validators[i];
+
+                if (aggregateWeight > 0) {
+                    aggregatePublicKey = BLST.addG1(aggregatePublicKey, validator.blsPublicKey);
+                    aggregateWeight += validator.weight;
+                } else {
+                    aggregatePublicKey = validator.blsPublicKey;
+                    aggregateWeight = validator.weight;
+                }
+            }
+        }
+        return (aggregatePublicKey, aggregateWeight);
+    }
+
+    // Verifies that quorumNum * totalWeight <= quorumDen * signatureWeight
+    function verifyWeight(
+        uint64 signatureWeight,
+        uint64 totalWeight
+    ) internal pure returns (bool) {
+        uint256 scaledTotalWeight = QUORUM_NUM * uint256(totalWeight);
+        uint256 scaledSignatureWeight = QUORUM_DEN * uint256(signatureWeight);
+        return scaledTotalWeight <= scaledSignatureWeight;
+    }
+
+    function verifyICMMessage(
+        ICMMessage memory message,
+        uint32 avalancheNetworkID,
+        bytes32 avalancheBlockChainID,
+        ValidatorSet memory validatorSet
+    ) internal view  {
+        if (message.unsignedMessage.avalancheNetworkID != avalancheNetworkID) {
+            revert("Invalid avalanche network ID");
+        }
+
+        // TODO: Do we need to check the avalanche source blockchain ID?
+        // It's expected to be different in cases where the message is from the primary network.
+        // if (
+        //     message.unsignedMessage.avalancheSourceBlockchainID
+        //         != validatorSet.avalancheBlockchainID
+        // ) {
+        //     revert("Invalid avalanche source blockchain ID");
+        // }
+
+
+         if (
+            avalancheBlockChainID != validatorSet.avalancheBlockchainID
+         ) {
+             revert("Invalid avalanche source blockchain ID");
+         }
+
+        bool[] memory signers = bytesToBoolArray(message.signature.signers);
+        (bytes memory aggregatePublicKey, uint64 aggregateWeight) =
+            filterValidators(signers, validatorSet.validators);
+
+        if (!verifyWeight(aggregateWeight, validatorSet.totalWeight)) {
+            revert("Invalid weight");
+        }
```

Comment:
I might also rename `verifyWeight` to `verifySignatureWeight` or `verifySufficientWeight`. At first I thought we were comparing if the two are equal.

```suggestion
        if (!verifyWeight(aggregateWeight, validatorSet.totalWeight)) {
            revert("Insufficient weight");
        }
```